### PR TITLE
Fix Issue 23648 - replace uses of sprintf with snprintf in the compiler

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: compiler/src/dmd/tocsym.d [hunk 1]
```diff
@@ -566,9 +567,10 @@ Symbol *toThunkSymbol(FuncDeclaration fd, int offset)
     s.Sfunc.Fflags &= ~Finline;  // thunks are not real functions, don't inline them
 
     __gshared int tmpnum;
-    char[6 + tmpnum.sizeof * 3 + 1] name = void;
+    const nameLen = 6 + tmpnum.sizeof * 3 + 1;
+    char[nameLen] name = void;
 
-    const len = sprintf(name.ptr,"_THUNK%d",tmpnum++);
+    const len = snprintf(name.ptr,nameLen,"_THUNK%d",tmpnum++);
```
Comment (dkorpel) [inline compiler/src/dmd/tocsym.d:573]: Why not just this?
```suggestion
    const len = snprintf(name.ptr,name.length,"_THUNK%d",tmpnum++);
```

File: compiler/src/dmd/tokens.d [hunk 2]
```diff
@@ -946,16 +946,17 @@ nothrow:
...
     {
-        __gshared char[3 + 3 * floatvalue.sizeof + 1] buffer;
+        const bufflen = 3 + 3 * floatvalue.sizeof + 1;
+        __gshared char[bufflen] buffer;
         const(char)* p = &buffer[0];
         switch (value)
         {
         case TOK.int32Literal:
-            sprintf(&buffer[0], "%d", cast(int)intvalue);
+            snprintf(&buffer[0], bufflen, "%d", cast(int)intvalue);
```
Comment (dkorpel) [inline compiler/src/dmd/tokens.d:955]: Similarly
```suggestion
            snprintf(&buffer[0], buffer.length, "%d", cast(int)intvalue);
```

File: compiler/src/dmd/libmach.d [hunk 3]
```diff
@@ -512,32 +518,27 @@ extern (C++) void MachOmToHeader(MachLibHeader* h, MachObjModule* om)
     int nzeros = 8 - ((slen + 4) & 7);
     if (nzeros < 4)
         nzeros += 8; // emulate mysterious behavior of ar
-    size_t len = sprintf(h.object_name.ptr, "#1/%lld", cast(long)(slen + nzeros));
+    size_t len = snprintf(h.object_name.ptr, MACH_OBJECT_NAME_SIZE, "#1/%lld", cast(long)(slen + nzeros));
```
Comment (dkorpel) [inline compiler/src/dmd/libmach.d:521]: No need to add extra indirections with the enums.
```suggestion
    size_t len = snprintf(h.object_name.ptr, h.object_name.length, "#1/%lld", cast(long)(slen + nzeros));
```

File: compiler/src/dmd/tokens.d (hunk 2 above)
Comment (RazvanN7) [reply compiler/src/dmd/tokens.d:955]: yes, but here buffer.length is used more often so I wanted to avoid the indirection.

**Comment by RazvanN7** [comment]
@dkorpel I think that in this issue it's not that important that we are consistent on all call sites. Sometimes I used a variable whereas sometimes I just used array.length, I think that this PR is acceptable as is since future additions on snprintf might use one or the other depending on context.

File: compiler/src/dmd/tokens.d (hunk 2 above)
Comment (dkorpel) [reply compiler/src/dmd/tokens.d:955]: What indirection? It's a compile time constant.

**Review by dkorpel — APPROVED** [review]

File: compiler/src/dmd/tokens.d (hunk 2 above)
Comment (RazvanN7) [reply compiler/src/dmd/tokens.d:955]: Oh, yeah makes sense, I did not know that.

**Comment by ibuclaw** [comment]
This PR caused a regression https://issues.dlang.org/show_bug.cgi?id=23759

**Pull request merged**